### PR TITLE
Porting PRs from corefx.

### PR DESCRIPTION
**Pull request by StephenBonikowsky**
Porting 4 PRs from corefx in 3 separate commits.

### CODE REVIEW

**Comment by StephenBonikowsky** [comment]
@dagood your PR dotnet/corefx#7989 caught several cases where we were requesting invalid packages.
I was not able to use msbuild /t:UpdatePackageDependencyVersion to update the json files, the tool was throwing an error, I didn't investigate and just updated manually. However I would love to be able to use that functionality so will look into it at some point.

**Comment by dagood** [comment]
Ah, it looks like I updated buildtools after you grabbed the version to use here. There was a bug with `UpdatePackageDependencyVersion` that I fixed and updated the PR with a newer fixed buildtools. It should now be `1.0.25-prerelease-00322-04` instead of `1.0.25-prerelease-00321-01`, sorry about that.

**Comment by StephenBonikowsky** [comment]
No problem, updated with new buildtools version.

File: BuildToolsVersion.txt [hunk 1]
```diff
@@ -1 +1 @@
-1.0.25-prerelease-00319-03
\ No newline at end of file
+1.0.25-prerelease-00322-04
```
Comment (iamjasonp) [inline BuildToolsVersion.txt]: Oops, my bad. I updated the version to 00321 with a previous PR not realizing you also touched the file here. You'll need to rebase this before merge.

**Comment by iamjasonp** [comment]
@dotnet-bot test Innerloop OSX Debug Build and Test, test Innerloop OSX Release Build and Test

**Comment by iamjasonp** [comment]
@dotnet-bot test Innerloop OSX Debug Build and Test

File: build.proj [hunk 2]
```diff
@@ -76,9 +83,31 @@
...
+  <!-- Evaluate our test projects (in src\tests.builds) -->
+  <Target Name="BatchGenerateTestProjectJsons"
+          Condition="'$(BuildTestsAgainstPackages)' == 'true'" 
+          DependsOnTargets="FilterProjects"
+          BeforeTargets="RestorePackages">
+      <Message Importance="High" Text="[$([System.DateTime]::Now.ToString('HH:mm:ss.ff'))] Generating Test project.json's..." />
+      <MSBuild Targets="GenerateAllTestProjectJsons"
+               Projects="src\tests.builds" />
+      <Message Importance="High" Text="[$([System.DateTime]::Now.ToString('HH:mm:ss.ff'))] Generating Test project.json's...done" />
+  </Target>
```
Comment (iamjasonp) [inline build.proj]: Question only: this means that if I do build.cmd /p:BuildTestsAgainstPackages=true, then i'll also be testing just against whatever published packages are? 

It seems like we pulled in only part of this change - see https://github.com/dotnet/corefx/commit/c1ec94b74183f598513199cde0d799a514c0a0fb. Do we need to pull the changes in dir.props as well?

**Comment by iamjasonp** [comment]
Builds fine for me in TFS as well. LGTM

NIT: personally - I prefer comments for "porting PR #..." to reference the SHA instead (e.g., `dotnet/corefx@c1ec94b`). It's a little easier to cross reference the commits, and also allows me to take a look without referring to GitHub :smile: Not blocking, and no need to change this PR.

**Comment by iamjasonp** [comment]
@dotnet-bot test outerloop Windows_NT, test outerloop Ubuntu14.04, test outerloop OSX

File: BuildToolsVersion.txt (hunk 1 above)
Comment (roncain) [inline BuildToolsVersion.txt]: I notice CoreFx is still on 321, and it's always a little risky to be different than them.  We might want to coordinate with them or submit a PR to CoreFx to go in at the same time as we update to 322.

**Comment by roncain** [comment]
LGTM, subject to question from @iamjasonp about whether we need dir.props and question from me whether we should update CoreFx's BuiltToolsVersion.txt to be 322 to avoid being out of sync.

File: BuildToolsVersion.txt (hunk 1 above)
Comment (dagood) [inline BuildToolsVersion.txt]: `00322-04` is part of https://github.com/dotnet/corefx/pull/7989 which I'm merging when I get a clean Ubuntu Release CI run.

File: build.proj (hunk 2 above)
Comment (StephenBonikowsky) [inline build.proj]: Those changes are already in.

File: BuildToolsVersion.txt (hunk 1 above)
Comment (dagood) [inline BuildToolsVersion.txt]: This has been merged into corefx now.

Comment (StephenBonikowsky) [inline BuildToolsVersion.txt]: Thanks!

**Comment by StephenBonikowsky** [comment]
Rebased and merging.

**Pull request merged**